tests/README.md

```diff
@@ -0,0 +1,1510 @@
+
+	public function setUp() {
+		parent::setUp();
+
+		$this->subject = new Factory();
+	}
+
+	// ... other parts of the test class ...
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function test_creation_of_standard_button() {
+		$this->assertInstanceOf
+				( Standard::class
+				, $this->subject->standard("label", "http://www.ilias.de")
+				);
+	}
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function test_creation_of_primary_button() {
+		$this->assertInstanceOf
+					( Primary::class
+					, $this->subject->primary("label", "http://www.ilias.de")
+					);
+	}
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function test_creation_of_Shy_button() {
+		$this->assertInstanceOf
+					( Shy::class
+					, $this->subject->shy("label", "http://www.ilias.de")
+					);
+	}
+
+	/**
+	 * @test
+	 * @small
+	 */
+	public function test_creation_of_close_button() {
+		$this->assertInstanceOf
+					( Close::class
+					, $this->subject->close("label", "http://www.ilias.de")
+					);
+	}
+}
+```
+
+
+<a name="useless-test--generic-naming"></a>
+#### Useless test / Generic naming
+Unit tests should always have an assertion of the result, because of that PHPUnit 6 started to mark such
+tests as useless. Useless test are always threaded as failed. Furthermore, the test name *test_button_label_or_glyph_only*
+is not really telling whats exactly tested.
+
+Another aspect of the whole *ButtonTest* class is that a factory is used to create concrete instances of
+the Buttons. But the factory nor the concrete subclasses of the Button class is a test subject here only the
+Button class itself. Of course to test all button instances is not really effective. However, the button can
+also be created with the help of mockery which subclasses the button dynamically within the tests. This allows
+to test the Button class in a dedicated way. If logic is added to one of the specific implementations only
+this part has to be tested within the test class of the specific button subclass for example the primary button.
+
+
+```php
+<?php
+class ButtonTest extends ILIAS_UI_TestBase {
+
+	public function getButtonFactory() {
+			return new \ILIAS\UI\Implementation\Component\Button\Factory();
+	}
+
+	public function button_type_provider() {
+			return array
+				( array("standard")
+				, array("primary")
+				, array("shy")
+				, array("tag")
+				);
+	}
+
+	/**
+	 * @dataProvider button_type_provider
+	 */
+	public function test_button_label_or_glyph_only($factory_method) {
+		$f = $this->getButtonFactory();
+		try {
+			$f->$factory_method($this, "http://www.ilias.de");
+			$this->assertFalse("This should not happen");
+		}
+		catch (\InvalidArgumentException $e) {}
+	}
+}
+```
+
+<a name="example-solution-1"></a>
+##### Example solution
+The proposed solution of this example would be to remove the test because the trait should be tested in a separate
+class dedicated to the trait. However, if the author decides to test the class which use the trait in stead of the trait itself
```

Comment:
use**s**

Further I guess, you mean the ComponentHelper trait, correct? However, the point here is not to test, whether this trait work or not, the point here is to test, whether is is applied correctly in the above instances of button. This is critical. E.g. what if somebody decides to override the constructor in primary and misses the necessary argument checks?

Reply:
Yes, the reason why I mentioned the trait was the assumption that no other logic was around except the type checks which are located in the trait it self. However your goal was the verification of the parameter validation. I will change the the description of the example solution accordingly.

I see your point, as far as I understood all buttons behave like the base button class. That's why I suggested to test the base class instead of all the concrete classes. In addition class specific parts could be handled within a separate test class for example PrimaryTest which tests all the class specific parts, for example the overwritten constructor.